les/catalyst/api.go

```diff
@@ -66,124 +78,222 @@ func NewConsensusAPI(les *les.LightEthereum) *ConsensusAPI {
+			return beacon.ForkChoiceResponse{PayloadStatus: beacon.INVALID_TERMINAL_BLOCK, PayloadID: nil}, nil
+		}
+	}
+	valid := func(id *beacon.PayloadID) beacon.ForkChoiceResponse {
+		return beacon.ForkChoiceResponse{
+			PayloadStatus: beacon.PayloadStatusV1{Status: beacon.VALID, LatestValidHash: &update.HeadBlockHash},
+			PayloadID:     id,
+		}
+	}
+	if rawdb.ReadCanonicalHash(api.les.ApiBackend.ChainDb(), header.Number.Uint64()) != update.HeadBlockHash {
+		// Block is not canonical, set head.
+		if err := api.les.BlockChain().SetCanonical(header); err != nil {
+			return beacon.ForkChoiceResponse{PayloadStatus: beacon.PayloadStatusV1{Status: beacon.INVALID, LatestValidHash: nil}}, err
+		}
+	} else if api.les.BlockChain().CurrentHeader().Hash() == update.HeadBlockHash {
+		// If the specified head matches with our local head, do nothing and keep
+		// generating the payload. It's a special corner case that a few slots are
+		// missing and we are requested to generate the payload in slot.
+	} else {
+		// If the head block is already in our canonical chain, the beacon client is
+		// probably resyncing. Ignore the update.
+		log.Info("Ignoring beacon update to old head", "number", header.Number.Uint64(), "hash", update.HeadBlockHash, "age", common.PrettyAge(time.Unix(int64(header.Time), 0)), "have", api.les.BlockChain().CurrentHeader().Number.Uint64())
+		return valid(nil), nil
+	}
+
+	// If the beacon client also advertised a finalized block, mark the local
+	// chain final and completely in PoS mode.
+	if update.FinalizedBlockHash != (common.Hash{}) {
+		if merger := api.les.Merger(); !merger.PoSFinalized() {
+			merger.FinalizePoS()
 		}
+		// If the finalized block is not in our canonical tree, somethings wrong
+		finalBlock, err := api.les.BlockChain().GetBlockByHash(ctx, update.FinalizedBlockHash)
+		if finalBlock == nil || err != nil {
+			log.Warn("Final block not available in database", "hash", update.FinalizedBlockHash)
+			return beacon.STATUS_INVALID, beacon.InvalidForkChoiceState.With(errors.New("final block not available in database"))
+		} else if rawdb.ReadCanonicalHash(api.les.ApiBackend.ChainDb(), finalBlock.NumberU64()) != update.FinalizedBlockHash {
+			log.Warn("Final block not in canonical chain", "number", header.Number.Uint64(), "hash", update.HeadBlockHash)
+			return beacon.STATUS_INVALID, beacon.InvalidForkChoiceState.With(errors.New("final block not in canonical chain"))
+		}
+		// Set the finalized block
+		// TODO (MariusVanDerWijden): Enable this once Finalized is implemented in LES
+		// api.les.BlockChain().SetFinalized(finalBlock)
 	}
-	// SetHead
-	if err := api.setCanonical(heads.HeadBlockHash); err != nil {
-		return beacon.STATUS_INVALID, err
+	// Check if the safe block hash is in our canonical tree, if not somethings wrong
+	if update.SafeBlockHash != (common.Hash{}) {
+		safeBlock, err := api.les.BlockChain().GetBlockByHash(ctx, update.SafeBlockHash)
```

Comment:
Please don't use GetBlockByHash which will retrieve the block from LES server if block is non-existent.